### PR TITLE
Added MapFeederForkJoin and a StreamMapper flag to use it. Ref T53505

### PR DESCRIPTION
**Pull request by yuesong**
In my local synthetic tests that simulates a few corner cases, the new MapFeederForkJoin achieved better bundle processing throughput than the current impl. In a test where I copied real source data and job config, it did the same.

### CODE REVIEW

**Comment by cburroughs** [comment]
:heart: forkjoin

**Comment by tea-dragon** [comment]
gotta give you points for craftyness, and forkjoin is one of those weird places where it is implemented so much more efficiently than a lot of its friends that even if you are neither forking nor joining a single time, sometimes (it at least feels) more efficient.

without comment on overall design or efficiency, I have a couple low hanging fruit guys though. i'll just inline them

File: hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    private final Semaphore enqueuePermits;
+
+    private long start = System.currentTimeMillis();
+    private long totalBundles = 0;
+
+    public MapFeederForkJoin(StreamMapper task, TaskDataSource source, int parallelism) {
+        this.source = source;
+        this.task = task;
+        this.parallelism = parallelism;
+        this.mapperPool = new ForkJoinPool(parallelism);
```
Comment (tea-dragon) [inline hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java]: you can hand the constructor an uncaught exception handler to save on a lot of try/finally

File: hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    private boolean fillBuffer() {
+        boolean status = false;
+        try {
+            enqueuePermits.acquire();
+            Bundle p = source.next();
+            if (p == null) {
+                log.info("exiting on null bundle from {}", source);
+            } else {
+                totalBundles++;
+                mapperPool.submit(new MapperTask(p));
```
Comment (tea-dragon) [inline hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java]: do we need to both create a task and then have the task do its own post-complete accounting? with the uncaught exception handler and a listener on the future, it can just be like a method reference

File: hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            log.info("exiting on premature stream termination");
+        } catch (InterruptedException ignored) {
+            Thread.currentThread().interrupt();
+            status = true;
+        }
+        return status;
+    }
+
+    private void joinProcessors() {
+        mapperPool.shutdown();
```
Comment (tea-dragon) [inline hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java]: have to actually then wait for termination

File: hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    /**
+     * Immediately halting is the best way we have of ensuring errors are reported without a significantly
+     * more involved system.
+     */
+    private void handleUncaughtThrowable(Throwable t) {
+        if (errored.compareAndSet(false, true)) {
+            log.error("unrecoverable error in task feeder or one of its mapper threads. immediately halting jvm", t);
+            Runtime.getRuntime().halt(1);
```
Comment (tea-dragon) [inline hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java]: new method in basis will replace this

File: hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java (hunk 3 above)
Comment (mspiegel) [inline hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java]: Whoops. And we probably want to wait for a finite amount of time and error the job if the wait does not complete successfully.

**Comment by tea-dragon** [comment]
even if the tests are admittedly synthetic, it would be cool to share them

File: hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java (hunk 4 above)
Comment (yuesong) [inline hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java]: what method?

Comment (tea-dragon) [inline hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java]: it's in the latest snapshot, and I mean specifically the logic contained in the handler (the Runtime.getRuntime().halt(1) call). The current namespacing is `Shutdown.exit(int)`

Comment (tea-dragon) [inline hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java]: (I originally used halt there because Runtime.exit(int) is so unreliable and crazy)

**Comment by yuesong** [comment]
As for the synthetic tests, they are simple jobs using constant source (which introduces imbalanced work sharding). I created three test cases, for each of which I run three jobs with identical config except that: one uses the original MapFeeder, one uses the forkjoin MapFeeder (`useForkJoinMapFeeder=true`), and one uses the original MapFeeder with work stealing (`// -Dtask.worksteal=true`), then I just compare the final throughputs. The test configs (using the original MapFeeder) are

- full speed (no filtering, empty output)
```
// -Dtask.worksteal=false

include "extra-filters"

source.const{
    bundles: [
        // thread 0
        {SHARD:"foo", A:a}
        // thread 1
        {SHARD:"bar", B:b}
        {SHARD:"bar", B:b}
        {SHARD:"bar", B:b}
        {SHARD:"bar", B:b}
    ]
    shardField:"SHARD"
    repeat:10000000
}

map.fields:[SHARD, A, B]
map.filterOut:[]

output.empty {}
```
- tree output (no filtering, simple tree output)
```
// -Dtask.worksteal=false

include "extra-filters"

source.const{
    bundles: [
        // thread 0
        {SHARD:"foo", A:a}
        // thread 1
        {SHARD:"bar", B:b}
        {SHARD:"bar", B:b}
        {SHARD:"bar", B:b}
        {SHARD:"bar", B:b}
    ]
    shardField:"SHARD"
    repeat:5000000
}

map.fields:[SHARD, A, B]
map.filterOut:[]

output.tree.root=[
    {const:root}
    {field:SHARD}
]
```
- slow filter (some bundles take longer than others) - this is the main use case we want to optimize
```
// -Dtask.worksteal=false

include "extra-filters"

source.const{
    bundles: [
        // thread 0
        {SHARD:"foo", A:a}
        // thread 1
        {SHARD:"bar", B:b}
        {SHARD:"bar", B:b}
        {SHARD:"bar", B:b}
        {SHARD:"bar", B:b}
    ]
    shardField:"SHARD"
    repeat:10000
}

map.fields:[SHARD, A, B]
map.filterOut:[
    {from:SHARD, to:HASH, hash {}}
    {log:sharding, maxBundles:10}
    {safely:[
        {has:A}
        {sleep.duration:2}
    ]}
    {safely:[
        {has:B}
        {sleep.duration:1}
    ]}
]

output.empty {}
```

**Comment by mspiegel** [comment]
It may be out of scope for this pull request but we can remove the shardField from the TaskDataSource. Unless there is some secret ThreadLocal variables there is no purpose for having a worker thread process a specific partition of the data.

**Comment by tea-dragon** [comment]
while I am by no means in favor of the shard field, there are regrettably a non-trivial number of possible uses. I haven't seen anyone try these or anything but:

- manipulation of the queues as seen in the tests (to somehow influence priority or downstream behavior)
- some filters have transient, thread-safe logic that can be keyed on bundles. For instance, all AutoField objects would prefer to have bundles of the same format from the same thread -- they don't use actual thread locals, but they use non-volatile fields that inherently benefit from cpu caching
- other filters have transient or even non-transient (builders et al) logic that is sometimes keyed on certain bundle fields

the most obvious one though is outputs. They always have to handle concurrent threads, and usually have side effects that care about more than their current bundle (and almost always including the bundles processed by other threads). You can imagine how sharding on a certain field that partitions a tree fairly well might improve throughput.

**Comment by tea-dragon** [comment]
As for the test cases, it would be relatively easy to make them into test classes right? Even just as a gist or something. These cases are nice, but they do seem a little biased to wards cases that are better for this implementation, and the sharding adds quite a few unknowns of its own.

What if the source is being the bottleneck? What is the change in CPU/memory overhead? For instance, one possibly worse case I can imagine is heavy cpu contention on the system and especially in the source (expensive decompression/deser perhaps). 

My gut says those costs are much less than eg. getting stuck even if we were to run into that case much more frequently (thus why I made the ticket), but performance can be a very unintuitive thing sometimes and usually worth considering all the angles.

**Comment by yuesong** [comment]
I'm not familiar with how sharding bundles to different worker threads can affect tree output efficiency? Regardless of shardFiled, as far as the output is concerned, the bundles are being added in somewhat random order with respect to the sharding, right? For implementation like the forkjoin one, it isn't applicable. I guess I just don't see people understanding/using the shardField, so...

The reason for the lack of a test case where source is the bottleneck is really that I can't easily introduce slowness to a source. That reminds me of my plan of using a real source for this...

**Comment by yuesong** [comment]
I just did a quick test of using a mesh2 source with empty map and output (so 97% of time is spent in source). Original MapFeeder had the lowest throughput, while MapFeederForkJoin and the original with work stealing performed the same.

**Comment by tea-dragon** [comment]
To be clear, I was specifically worried about a case where system CPU was constrained. Unless you had a million threads and were monitoring the overall system load, I don't see how this way measured.
